include/deal.II/base/tensor_base.h

```diff
@@ -1388,163 +1298,111 @@ struct ProductType<Tensor<1,dim,T>,U>
  * you would get if you multiplied an individual component of the input tensor
  * by the scalar factor.
  *
- * @relates Tensor<1,dim,Number>
+ * @relates Tensor
  * @relates EnableIfScalar
  */
-template <int dim, typename Number, typename OtherNumber>
+template <int rank, int dim,
+          typename Number,
+          typename OtherNumber,
+          typename = typename EnableIfScalar<OtherNumber>::type>
 inline
-Tensor<1,dim,typename ProductType<Number,typename EnableIfScalar<OtherNumber>::type>::type>
-operator * (const Tensor<1,dim,Number> &t,
-            const OtherNumber           factor)
+Tensor<rank,dim,typename ProductType<Number, OtherNumber>::type>
+operator * (const Tensor<rank,dim,Number> &t,
+            const OtherNumber              factor)
 {
-  // form the product. we have to convert the two factors into the final
-  // type via explicit casts because, for awkward reasons, the C++
-  // standard committee saw it fit to not define an
-  //   operator*(float,std::complex<double>)
-  // (as well as with switched arguments and double<->float).
-  typedef typename ProductType<Number,OtherNumber>::type product_type;
-  Tensor<1,dim,product_type> tt (false);
+  // recurse over the base objects
+  Tensor<rank,dim,typename ProductType<Number,OtherNumber>::type> tt;
   for (unsigned int d=0; d<dim; ++d)
-    tt[d] = product_type(t[d]) * product_type(factor);
+    tt[d] = t[d] * factor;
```

Comment:
No, I ran into this earlier this year with GCC when helping to make the data type in `FEValues` generic (based on @davydden's patches). It's not a libc++ problem alone.